packages/rum-react/src/get-apm-routes.js

```diff
@@ -0,0 +1,142 @@
+import {
+  Routes,
+  useLocation,
+  useNavigationType,
+  matchRoutes,
+  createRoutesFromChildren
+} from 'react-router-dom'
+import hoistNonReactStatics from 'hoist-non-react-statics'
+
+function getApmRoutes(apm) {
+  function ApmRoutes(props) {
+    const location = useLocation()
+    const navigationType = useNavigationType()
+
+    React.useLayoutEffect(
+      () => {
+        const routes = createRoutesFromChildren(props.children)
+        handleRouteChange(apm, routes, location)
+      },
+      // query param changes should not trigger this effect
+      // that's why we use `location.pathname` instead of `location`
+      [location.pathname, navigationType, props.children]
+    )
+
+    useFinishTransactionEffect(apm.getCurrentTransaction())
+
+    return <Routes {...props} />
+  }
+
+  hoistNonReactStatics(ApmRoutes, Routes)
+
+  return ApmRoutes
+}
+
+function handleRouteChange(apm, routes, location, navigationType) {
+  if (!apm.isActive()) {
+    return
+  }
+
+  // one of the cases when REPLACE gets triggered is when someone clicks on a link that leads to the current page
+  // we don't want to handle this case
+  if (navigationType === 'REPLACE') {
+    return
+  }
+
+  // just for navigationType PUSH and POP
+  const name = getRouteFromLocation(routes, location)
+  apm.startTransaction(name, 'route-change', {
+    managed: true,
+    canReuse: true
+  })
+}
+
+export function getRouteFromLocation(routes, location) {
+  if (!routes || routes.length === 0) {
+    return location.pathname
+  }
+
+  // There are cases where we can have more than one matchedRoute.
+  // For instance, when using Nested Routes with a configuration such as `/user`, `/user/profile` and `/user/account`
+  // if a user clicks on the link '/user/profile'
+  // matchRoutes will return two results, a `route.path` with `/user` and the another one with `profile`
+  const branches = matchRoutes(routes, location)
+  let builtPath = ''
+  if (branches) {
+    for (let index = 0; index < branches.length; index++) {
+      const branch = branches[index]
+      const route = branch.route
+      if (route) {
+        if (route.index) {
+          // this cover the cases where an index route has been defined. They don't have a `route.path` property
+          // E.g. <Route index element={<Home />} />
+          return branch.pathname
+        }
+
+        let path = route.path
+        if (path) {
+          // make sure we don't add an extra slash when concatenating
+          const hasSlash =
+            path[0] === '/' || builtPath[builtPath.length - 1] === '/'
+          path = hasSlash ? path : `/${path}`
+          builtPath += path
+
+          // E.g `to` property from `<Link>` component matches to window.location.pathname
+          if (branch.pathname === location.pathname) {
+            // make sure wildcard routes are also handled
+            const isWildcardRoute = builtPath.slice(-2) === '/*'
```

Comment:
Does wildcard paths always end in this way? What about `path="/*"` ?

Reply:
> A * wildcard may be used only at the end of a path, not in the middle.

From the docs "upgrading from v5"
https://reactrouter.com/en/main/upgrading/v5#note-on-route-path-patterns